File: Shared/ChartView.m
//
//  ChartView.m
//  StockList Demo for iOS
//
// Copyright 2013 Weswit Srl
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#import "ChartView.h"

#import <CoreText/CoreText.h>

#define LEFT_AXIS_MARGIN            (40.0)
#define BOTTOM_AXIS_MARGIN          (20.0)
#define RIGHT_MARGIN                (40.0)

#define CHART_BACKGROUND_COLOR      ([UIColor colorWithRed:0.7176 green:0.8431 blue:0.9647 alpha:1.0000])
#define AXIS_LINE_COLOR             ([UIColor darkGrayColor])
#define AXIS_LABEL_COLOR            ([UIColor darkGrayColor])
#define INTEGER_LINE_COLOR          ([UIColor darkGrayColor])
#define INTEGER_LABEL_COLOR         ([UIColor darkGrayColor])
#define VALUE_LINE_COLOR            ([UIColor blueColor])
#define VALUE_LABEL_COLOR           ([UIColor blueColor])

#define AXIS_LINE_WIDTH              (2.0)
#define INTEGER_LINE_WIDTH           (1.0)
#define VALUE_LINE_WIDTH             (2.0)

#define FONT_SIZE                   (11.0)


@implementation ChartView


#pragma mark -
#pragma mark Initialization

- (id) initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {

        // Initialization
		_data= [[NSMutableArray alloc] init];
		
		self.clearsContextBeforeDrawing= YES;
		self.clipsToBounds= YES;
    }
	
    return self;
}


#pragma mark -
#pragma mark Data management

- (void) addValue:(float)value withTime:(NSTimeInterval)time {
	@synchronized (self) {
	
		// Add value, keeping the array sorted
		NSValue *point= [NSValue valueWithCGPoint:CGPointMake(time, value)];

		NSUInteger index= [_data indexOfObject:point
								 inSortedRange:NSMakeRange(0, [_data count])
									   options:NSBinarySearchingLastEqual | NSBinarySearchingInsertionIndex
							   usingComparator:^NSComparisonResult(id obj1, id obj2) {
								   CGPoint point1= [(NSValue *) obj1 CGPointValue];
								   CGPoint point2= [(NSValue *) obj2 CGPointValue];
							
								   if (point1.x < point2.x)
									   return NSOrderedAscending;
								   else if (point1.x > point2.x)
									   return NSOrderedDescending;
								   else
									   return NSOrderedSame;
							   }];
		
		[_data insertObject:[NSValue valueWithCGPoint:CGPointMake(time, value)] atIndex:index];
		
		// Strip out values out of time range
		NSValue *begin= [NSValue valueWithCGPoint:CGPointMake(_begin, 0.0)];
		
		index= [_data indexOfObject:begin
					  inSortedRange:NSMakeRange(0, [_data count])
							options:NSBinarySearchingFirstEqual | NSBinarySearchingInsertionIndex
					usingComparator:^NSComparisonResult(id obj1, id obj2) {
						CGPoint point1= [(NSValue *) obj1 CGPointValue];
						CGPoint point2= [(NSValue *) obj2 CGPointValue];
						
						if (point1.x < point2.x)
							return NSOrderedAscending;
						else if (point1.x > point2.x)
							return NSOrderedDescending;
						else
							return NSOrderedSame;
					}];
		
		if (index >= 1)
			[_data removeObjectsAtIndexes:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, index)]];
		
		NSValue *end= [NSValue valueWithCGPoint:CGPointMake(_end, 0.0)];
		
		index= [_data indexOfObject:end
					  inSortedRange:NSMakeRange(0, [_data count])
							options:NSBinarySearchingLastEqual | NSBinarySearchingInsertionIndex
					usingComparator:^NSComparisonResult(id obj1, id obj2) {
						CGPoint point1= [(NSValue *) obj1 CGPointValue];
						CGPoint point2= [(NSValue *) obj2 CGPointValue];
						
						if (point1.x < point2.x)
							return NSOrderedAscending;
						else if (point1.x > point2.x)
							return NSOrderedDescending;
						else
							return NSOrderedSame;
					}];
		
		if (index < [_data count])
			[_data removeObjectsAtIndexes:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(index, [_data count] -index)]];
	}
	
	[self setNeedsDisplay];
}

- (void) clearValues {
	@synchronized (self) {
		[_data removeAllObjects];
	}
	
	[self setNeedsDisplay];
}


#pragma mark -
#pragma mark Coordinates managements

- (CGPoint) valueAtPoint:(CGPoint)point {
	float x= point.x;
	float y= point.y;
	
	CGSize size= self.frame.size;
	
	float propX= (x - LEFT_AXIS_MARGIN) / (size.width - LEFT_AXIS_MARGIN - RIGHT_MARGIN);
	float propY= (size.height - BOTTOM_AXIS_MARGIN - y) / (size.height - BOTTOM_AXIS_MARGIN);
	
	float time= _begin + propX * (_end - _begin);
	float value= _min + propY * (_max - _min);

	return CGPointMake(time, value);
}

- (CGPoint) pointForValue:(CGPoint)point {
	NSTimeInterval time= point.x;
	float value= point.y;
	
	float deltaX= time - _begin;
	float deltaY= value - _min;
	
	float propX= deltaX / (_end - _begin);
	float propY= deltaY / (_max - _min);
	
	CGSize size= self.frame.size;
	
	float x= LEFT_AXIS_MARGIN + ((size.width - LEFT_AXIS_MARGIN - RIGHT_MARGIN) * propX);
	float y= (size.height - BOTTOM_AXIS_MARGIN) - ((size.height - BOTTOM_AXIS_MARGIN) * propY);
	
	if (x < LEFT_AXIS_MARGIN)
		x= LEFT_AXIS_MARGIN;
	
	if (x >= size.width)
		x= size.width;
	
	if (y < 0.0)
		y= 0.0;
	
	if (y >= (size.height - BOTTOM_AXIS_MARGIN))
		y= (size.height - BOTTOM_AXIS_MARGIN);
	
	return CGPointMake(x, y);
}


#pragma mark -
#pragma mark Properties

@dynamic min;

- (float) min {
	return _min;
}

- (void) setMin:(float)min {
	@synchronized (self) {
		_min= min;
	}
	
	[self setNeedsDisplay];
}

@dynamic max;

- (float) max {
	return _max;
}

- (void) setMax:(float)max {
	@synchronized (self) {
		_max= max;
	}
	
	[self setNeedsDisplay];
}

@dynamic begin;

- (NSTimeInterval) begin {
	return _begin;
}

- (void) setBegin:(NSTimeInterval)begin {
	@synchronized (self) {
		_begin= begin;
	}
	
	[self setNeedsDisplay];
}

@dynamic end;

- (NSTimeInterval) end {
	return _end;
}

- (void) setEnd:(NSTimeInterval)end {
	@synchronized (self) {
		_end= end;
	}
	
	[self setNeedsDisplay];
}


#pragma mark -
#pragma mark Drawing

- (void) drawRect:(CGRect)rect {
	[super drawRect:rect];
	
	CGSize size= self.frame.size;
	CGContextRef context= UIGraphicsGetCurrentContext();
	
	CTFontRef sysFont= CTFontCreateUIFontForLanguage(kCTFontSystemFontType, FONT_SIZE, NULL);
	CTFontRef sysBoldFont= CTFontCreateUIFontForLanguage(kCTFontUIFontEmphasizedSystem, FONT_SIZE, NULL);
	
	CGAffineTransform transform= CGAffineTransformMakeTranslation(0.0, size.height);
	CGAffineTransform transform2= CGAffineTransformScale(transform, 1.0, -1.0);
	CGContextSetTextMatrix(context, transform2);
	
	@synchronized (self) {
		
		// Draw the chart background
		CGContextSetFillColorWithColor(context, CHART_BACKGROUND_COLOR.CGColor);
		CGContextFillRect(context, CGRectMake(LEFT_AXIS_MARGIN, 0.0, size.width - LEFT_AXIS_MARGIN, size.height - BOTTOM_AXIS_MARGIN));

		// Draw the two axis
		CGContextSetStrokeColorWithColor(context, AXIS_LINE_COLOR.CGColor);
		CGContextSetLineWidth(context, AXIS_LINE_WIDTH);

		CGContextMoveToPoint(context, LEFT_AXIS_MARGIN, 0.0);
		CGContextAddLineToPoint(context, LEFT_AXIS_MARGIN, size.height);
		CGContextStrokePath(context);

		CGContextMoveToPoint(context, 0.0, size.height - BOTTOM_AXIS_MARGIN);
		CGContextAddLineToPoint(context, size.width, size.height - BOTTOM_AXIS_MARGIN);
		CGContextStrokePath(context);
		
		// Draw integer lines
		CGContextSetStrokeColorWithColor(context, INTEGER_LINE_COLOR.CGColor);
		CGContextSetLineWidth(context, INTEGER_LINE_WIDTH);
		CTLineRef line= nil;
		
		NSDictionary *attributesDict= [NSDictionary dictionaryWithObjectsAndKeys:
									   (__bridge id) sysFont, (id) kCTFontAttributeName,
									   INTEGER_LABEL_COLOR.CGColor, (id) kCTForegroundColorAttributeName,
									   nil];
		
		float dashPhase= 0.0;
		float dashLengths[] = { 10.0, 5.0 };
		CGContextSetLineDash(context, dashPhase, (const CGFloat *) dashLengths, 2);
		
		int current= ((int) (_min + 1.0));
		while (current < _max) {
			
			// Compute position
			CGPoint point= [self pointForValue:CGPointMake(0.0, current)];
			float y= point.y;
			
			CGContextMoveToPoint(context, LEFT_AXIS_MARGIN, y);
			CGContextAddLineToPoint(context, size.width, y);
			CGContextStrokePath(context);
			
			// Draw label
			if ((y > FONT_SIZE + 5.0) && (y < (size.height - BOTTOM_AXIS_MARGIN - FONT_SIZE - 5.0))) {
				NSString *intLabel= [NSString stringWithFormat:@"%7.2f", ((float) current)];
				NSAttributedString *intLabelAttr= [[NSAttributedString alloc] initWithString:intLabel attributes:attributesDict];
				CGContextSetTextPosition(context, 0.0, y);
				
				CTLineRef line= CTLineCreateWithAttributedString((CFAttributedStringRef) intLabelAttr);
				CTLineDraw(line, context);
				CFRelease(line);
			}
			
			current++;
		}
		
		// Draw the Y axis labels
		attributesDict= [NSDictionary dictionaryWithObjectsAndKeys:
						 (__bridge id) sysBoldFont, (id) kCTFontAttributeName,
						 AXIS_LABEL_COLOR.CGColor, (id) kCTForegroundColorAttributeName,
						 nil];
		
		float maxY= FONT_SIZE;
		NSString *maxLabel= [NSString stringWithFormat:@"%7.2f", _max];
		NSAttributedString *maxLabelAttr= [[NSAttributedString alloc] initWithString:maxLabel attributes:attributesDict];
		CGContextSetTextPosition(context, 0.0, maxY);
		line= CTLineCreateWithAttributedString((CFAttributedStringRef) maxLabelAttr);
		CTLineDraw(line, context);
		CFRelease(line);
		
		float minY= size.height - BOTTOM_AXIS_MARGIN - 5.0;
		NSString *minLabel= [NSString stringWithFormat:@"%7.2f", _min];
		NSAttributedString *minLabelAttr= [[NSAttributedString alloc] initWithString:minLabel attributes:attributesDict];
		CGContextSetTextPosition(context, 0.0, minY);
		line= CTLineCreateWithAttributedString((CFAttributedStringRef) minLabelAttr);
		CTLineDraw(line, context);
		CFRelease(line);
		
		// Draw the X axis labels
		NSDateFormatter *formatter= [[NSDateFormatter alloc] init];
		[formatter setDateFormat:@"HH:mm:ss"];
		
		NSString *beginLabel= [formatter stringFromDate:[NSDate dateWithTimeIntervalSinceReferenceDate:_begin]];
		NSAttributedString *beginLabelAttr= [[NSAttributedString alloc] initWithString:beginLabel attributes:attributesDict];
		CGContextSetTextPosition(context, LEFT_AXIS_MARGIN + 5.0, size.height - 5.0);
		line= CTLineCreateWithAttributedString((CFAttributedStringRef) beginLabelAttr);
		CTLineDraw(line, context);
		CFRelease(line);
		
		NSString *middleLabel= [formatter stringFromDate:[NSDate dateWithTimeIntervalSinceReferenceDate:(_begin + _end) / 2.0]];
		NSAttributedString *middleLabelAttr= [[NSAttributedString alloc] initWithString:middleLabel attributes:attributesDict];
		CGContextSetTextPosition(context, LEFT_AXIS_MARGIN + ((size.width - LEFT_AXIS_MARGIN) / 2.0) - (FONT_SIZE * 2.0), size.height - 5.0);
		line= CTLineCreateWithAttributedString((CFAttributedStringRef) middleLabelAttr);
		CTLineDraw(line, context);
		CFRelease(line);
		
		NSString *endLabel= [formatter stringFromDate:[NSDate dateWithTimeIntervalSinceReferenceDate:_end]];
		NSAttributedString *endLabelAttr= [[NSAttributedString alloc] initWithString:endLabel attributes:attributesDict];
		CGContextSetTextPosition(context, size.width - (FONT_SIZE * 4.0), size.height - 5.0);
		line= CTLineCreateWithAttributedString((CFAttributedStringRef) endLabelAttr);
		CTLineDraw(line, context);
		CFRelease(line);
		
		// Draw data
		CGContextSetStrokeColorWithColor(context, VALUE_LINE_COLOR.CGColor);
		CGContextSetLineWidth(context, VALUE_LINE_WIDTH);
		CGContextSetLineDash(context, 0.0, NULL, 0);

		float lastValue= 0.0;
		float lastY= 0.0;
		BOOL first= YES;

		for (NSValue *pointValue in _data) {
			CGPoint point= [self pointForValue:[pointValue CGPointValue]];
			float x= point.x;
			float y= point.y;
			
			lastValue= [pointValue CGPointValue].y;
			lastY= y;
			
			if (first) {
				CGContextMoveToPoint(context, x, y);
				first= NO;

				// Draw at least a dot if there's just one point
				if ([_data count] == 1)
					CGContextAddLineToPoint(context, x + VALUE_LINE_WIDTH, y);
			
			} else
				CGContextAddLineToPoint(context, x, y);
		}
		
		if (!first) {
			
			// Draw line and label if there's at least one point
			CGContextStrokePath(context);
		
			// Draw last point label
			attributesDict= [NSDictionary dictionaryWithObjectsAndKeys:
							 (__bridge id) sysBoldFont, (id) kCTFontAttributeName,
							 VALUE_LABEL_COLOR.CGColor, (id) kCTForegroundColorAttributeName,
							 nil];
			
			NSString *label= [NSString stringWithFormat:@"%7.2f", lastValue];
			NSAttributedString *labelAttr= [[NSAttributedString alloc] initWithString:label attributes:attributesDict];
			CGContextSetTextPosition(context, size.width - RIGHT_MARGIN, lastY);
			line= CTLineCreateWithAttributedString((CFAttributedStringRef) labelAttr);
			CTLineDraw(line, context);
			CFRelease(line);
		}
	}
	
	// Cleanup
	CFRelease(sysFont);
	CFRelease(sysBoldFont);
}


@end

File: Shared/ChartView.swift
//  Converted to Swift 5.4 by Swiftify v5.4.24488 - https://swiftify.com/
//
//  ChartView.swift
//  StockList Demo for iOS
//
// Copyright (c) Lightstreamer Srl
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import CoreText
import UIKit

class ChartView: UIView {
    let lockQueue = DispatchQueue(label: "lightstreamer.ChartView")
    var data: [AnyHashable]?





    // MARK: -
    // MARK: Properties

    private var _min: Float = 0.0
    @objc var min: Float {
        get {
            return _min
        }
        set(min) {
            lockQueue.sync {
                _min = min
            }

            setNeedsDisplay()
        }
    }

    private var _max: Float = 0.0
    @objc var max: Float {
        get {
            return _max
        }
        set(max) {
            lockQueue.sync {
                _max = max
            }

            setNeedsDisplay()
        }
    }

    private var _begin: TimeInterval = 0.0
    @objc var begin: TimeInterval {
        get {
            return _begin
        }
        set(begin) {
            lockQueue.sync {
                _begin = begin
            }

            setNeedsDisplay()
        }
    }

    private var _end: TimeInterval = 0.0
    @objc var end: TimeInterval {
        get {
            return _end
        }
        set(end) {
            lockQueue.sync {
                _end = end
            }

            setNeedsDisplay()
        }
    }

    // MARK: -
    // MARK: Initialization

    override init(frame: CGRect) {
        super.init(frame: frame)
        // Initialization
        data = []

        clearsContextBeforeDrawing = true
        clipsToBounds = true
    }

    // MARK: -
    // MARK: Data management
    @objc func addValue(_ value: Float, withTime time: TimeInterval) {
        lockQueue.sync {

            // Add value, keeping the array sorted
            let point = NSValue(cgPoint: CGPoint(x: CGFloat(time), y: CGFloat(value)))

            var index = (data as NSArray?)?.index(of: point, inSortedRange: NSRange(location: 0, length: data?.count ?? 0), options: [.lastEqual, .insertionIndex], usingComparator: { obj1, obj2 in
                let point1 = (obj1 as? NSValue)?.cgPointValue
                let point2 = (obj2 as? NSValue)?.cgPointValue

                if (point1?.x ?? 0.0) < (point2?.x ?? 0.0) {
                    return .orderedAscending
                } else if (point1?.x ?? 0.0) > (point2?.x ?? 0.0) {
                    return .orderedDescending
                } else {
                    return .orderedSame
                }
            }) ?? 0

            data?.insert(NSValue(cgPoint: CGPoint(x: CGFloat(time), y: CGFloat(value))), at: index)

            // Strip out values out of time range
            let begin = NSValue(cgPoint: CGPoint(x: CGFloat(self.begin), y: 0.0))

            index = (data as NSArray?)?.index(of: begin, inSortedRange: NSRange(location: 0, length: data?.count ?? 0), options: [.firstEqual, .insertionIndex], usingComparator: { obj1, obj2 in
                let point1 = (obj1 as? NSValue)?.cgPointValue
                let point2 = (obj2 as? NSValue)?.cgPointValue

                if (point1?.x ?? 0.0) < (point2?.x ?? 0.0) {
                    return .orderedAscending
                } else if (point1?.x ?? 0.0) > (point2?.x ?? 0.0) {
                    return .orderedDescending
                } else {
                    return .orderedSame
                }
            }) ?? 0

            if index >= 1 {
                for deletionIndex in NSIndexSet(indexesIn: NSRange(location: 0, length: index)).reversed() { data?.remove(at: deletionIndex) }
            }

            let end = NSValue(cgPoint: CGPoint(x: CGFloat(self.end), y: 0.0))

            index = (data as NSArray?)?.index(of: end, inSortedRange: NSRange(location: 0, length: data?.count ?? 0), options: [.lastEqual, .insertionIndex], usingComparator: { obj1, obj2 in
                let point1 = (obj1 as? NSValue)?.cgPointValue
                let point2 = (obj2 as? NSValue)?.cgPointValue

                if (point1?.x ?? 0.0) < (point2?.x ?? 0.0) {
                    return .orderedAscending
                } else if (point1?.x ?? 0.0) > (point2?.x ?? 0.0) {
                    return .orderedDescending
                } else {
                    return .orderedSame
                }
            }) ?? 0

            if index < (data?.count ?? 0) {
                for deletionIndex in NSIndexSet(indexesIn: NSRange(location: index, length: (data?.count ?? 0) - index)).reversed() { data?.remove(at: deletionIndex) }
            }
        }

        setNeedsDisplay()
    }

    @objc func clearValues() {
        lockQueue.sync {
            data?.removeAll()
        }

        setNeedsDisplay()
    }

    // MARK: -
    // MARK: Coordinates managements
    @objc func value(at point: CGPoint) -> CGPoint {
        let x = Float(point.x)
        let y = Float(point.y)

        let size = frame.size

        let propX = Float((Double(x) - LEFT_AXIS_MARGIN) / (Double(size.width) - LEFT_AXIS_MARGIN - RIGHT_MARGIN))
        let propY = Float((Double(size.height) - BOTTOM_AXIS_MARGIN - Double(y)) / (Double(size.height) - BOTTOM_AXIS_MARGIN))

        let time = Float(begin) + propX * Float((end - begin))
        let value = min + propY * (max - min)

        return CGPoint(x: CGFloat(time), y: CGFloat(value))
    }

    func point(forValue point: CGPoint) -> CGPoint {
        let time = TimeInterval(point.x)
        let value = Float(point.y)

        let deltaX = Float(time - begin)
        let deltaY = value - min

        let propX = deltaX / Float((end - begin))
        let propY = deltaY / (max - min)

        let size = frame.size

        var x = Float(LEFT_AXIS_MARGIN + ((Double(size.width) - LEFT_AXIS_MARGIN - RIGHT_MARGIN) * Double(propX)))
        var y = Float((Double(size.height) - BOTTOM_AXIS_MARGIN) - ((Double(size.height) - BOTTOM_AXIS_MARGIN) * Double(propY)))

        if Double(x) < LEFT_AXIS_MARGIN {
            x = Float(LEFT_AXIS_MARGIN)
        }

        if CGFloat(x) >= size.width {
            x = Float(size.width)
        }

        if y < 0.0 {
            y = 0.0
        }

        if Double(y) >= (Double(size.height) - BOTTOM_AXIS_MARGIN) {
            y = Float((Double(size.height) - BOTTOM_AXIS_MARGIN))
        }

        return CGPoint(x: CGFloat(x), y: CGFloat(y))
    }

    // MARK: -
    // MARK: Properties

    // MARK: -
    // MARK: Drawing

    override func draw(_ rect: CGRect) {
        super.draw(rect)

        let size = frame.size
        let context = UIGraphicsGetCurrentContext()

        let sysFont = CTFontCreateUIFontForLanguage(.kCTFontSystemFontType, CGFloat(FONT_SIZE), nil)
        let sysBoldFont = CTFontCreateUIFontForLanguage(.emphasizedSystem, CGFloat(FONT_SIZE), nil)

        let transform = CGAffineTransform(translationX: 0.0, y: size.height)
        let transform2 = transform.scaledBy(x: 1.0, y: -1.0)
        context?.textMatrix = transform2

        lockQueue.sync {

            // Draw the chart background
            context?.setFillColor(CHART_BACKGROUND_COLOR.cgColor)
            context?.fill(CGRect(x: CGFloat(LEFT_AXIS_MARGIN), y: 0.0, width: CGFloat(Double(size.width) - LEFT_AXIS_MARGIN), height: CGFloat(Double(size.height) - BOTTOM_AXIS_MARGIN)))

            // Draw the two axis
            context?.setStrokeColor(AXIS_LINE_COLOR.cgColor)
            context?.setLineWidth(CGFloat(AXIS_LINE_WIDTH))

            context?.move(to: CGPoint(x: CGFloat(LEFT_AXIS_MARGIN), y: 0.0))
            context?.addLine(to: CGPoint(x: CGFloat(LEFT_AXIS_MARGIN), y: size.height))
            context?.strokePath()

            context?.move(to: CGPoint(x: 0.0, y: CGFloat(Double(size.height) - BOTTOM_AXIS_MARGIN)))
            context?.addLine(to: CGPoint(x: size.width, y: CGFloat(Double(size.height) - BOTTOM_AXIS_MARGIN)))
            context?.strokePath()

            // Draw integer lines
            context?.setStrokeColor(INTEGER_LINE_COLOR.cgColor)
            context?.setLineWidth(CGFloat(INTEGER_LINE_WIDTH))
            var line: CTLine? = nil

            var attributesDict: [AnyHashable : Any]? = nil
            if let sysFont = sysFont {
                attributesDict = [
                    kCTFontAttributeName : sysFont,
                    kCTForegroundColorAttributeName : INTEGER_LABEL_COLOR.cgColor
                ]
            }

            let dashPhase: Float = 0.0
            let dashLengths = [CGFloat(10.0), CGFloat(5.0)]
            context?.setLineDash(phase: CGFloat(dashPhase), lengths: dashLengths)

            var current = Int(min + 1.0)
            while Float(current) < max {

                // Compute position
                let point = self.point(forValue: CGPoint(x: 0.0, y: CGFloat(current)))
                let y = Float(point.y)

                context?.move(to: CGPoint(x: CGFloat(LEFT_AXIS_MARGIN), y: CGFloat(y)))
                context?.addLine(to: CGPoint(x: size.width, y: CGFloat(y)))
                context?.strokePath()

                // Draw label
                if (Double(y) > FONT_SIZE + 5.0) && (Double(y) < (Double(size.height) - BOTTOM_AXIS_MARGIN - FONT_SIZE - 5.0)) {
                    let intLabel = String(format: "%7.2f", Float(current))
                    let intLabelAttr = NSAttributedString(string: intLabel, attributes: attributesDict as? [NSAttributedString.Key : Any])
                    context?.textPosition = CGPoint(x: 0.0, y: CGFloat(y))

                    let line = CTLineCreateWithAttributedString(intLabelAttr as CFAttributedString)
                    if let context = context {
                        CTLineDraw(line, context)
                    }
                }

                current += 1
            }

            // Draw the Y axis labels
            if let sysBoldFont = sysBoldFont {
                attributesDict = [
                    kCTFontAttributeName : sysBoldFont,
                    kCTForegroundColorAttributeName : AXIS_LABEL_COLOR.cgColor
                ]
            }

            let maxY = Float(FONT_SIZE)
            let maxLabel = String(format: "%7.2f", max)
            let maxLabelAttr = NSAttributedString(string: maxLabel, attributes: attributesDict as? [NSAttributedString.Key : Any])
            context?.textPosition = CGPoint(x: 0.0, y: CGFloat(maxY))
            line = CTLineCreateWithAttributedString(maxLabelAttr as CFAttributedString)
            if let context = context {
                CTLineDraw(line!, context)
            }

            let minY = Float(Double(size.height) - BOTTOM_AXIS_MARGIN - 5.0)
            let minLabel = String(format: "%7.2f", min)
            let minLabelAttr = NSAttributedString(string: minLabel, attributes: attributesDict as? [NSAttributedString.Key : Any])
            context?.textPosition = CGPoint(x: 0.0, y: CGFloat(minY))
            line = CTLineCreateWithAttributedString(minLabelAttr as CFAttributedString)
            if let context = context {
                CTLineDraw(line!, context)
            }

            // Draw the X axis labels
            let formatter = DateFormatter()
            formatter.dateFormat = "HH:mm:ss"

            let beginLabel = formatter.string(from: Date(timeIntervalSinceReferenceDate: begin))
            let beginLabelAttr = NSAttributedString(string: beginLabel, attributes: attributesDict as? [NSAttributedString.Key : Any])
            context?.textPosition = CGPoint(x: CGFloat(LEFT_AXIS_MARGIN + 5.0), y: size.height - 5.0)
            line = CTLineCreateWithAttributedString(beginLabelAttr as CFAttributedString)
            if let context = context {
                CTLineDraw(line!, context)
            }

            let middleLabel = formatter.string(from: Date(timeIntervalSinceReferenceDate: (begin + end) / 2.0))
            let middleLabelAttr = NSAttributedString(string: middleLabel, attributes: attributesDict as? [NSAttributedString.Key : Any])
            context?.textPosition = CGPoint(x: CGFloat(LEFT_AXIS_MARGIN + ((Double(size.width) - LEFT_AXIS_MARGIN) / 2.0) - Double((middleLabelAttr.size().width / 2.0))), y: size.height - 5.0)
            line = CTLineCreateWithAttributedString(middleLabelAttr as CFAttributedString)
            if let context = context {
                CTLineDraw(line!, context)
            }

            let endLabel = formatter.string(from: Date(timeIntervalSinceReferenceDate: end))
            let endLabelAttr = NSAttributedString(string: endLabel, attributes: attributesDict as? [NSAttributedString.Key : Any])
            context?.textPosition = CGPoint(x: size.width - endLabelAttr.size().width, y: size.height - 5.0)
            line = CTLineCreateWithAttributedString(endLabelAttr as CFAttributedString)
            if let context = context {
                CTLineDraw(line!, context)
            }

            // Draw data
            context?.setStrokeColor(VALUE_LINE_COLOR.cgColor)
            context?.setLineWidth(CGFloat(VALUE_LINE_WIDTH))
            context?.setLineDash(phase: 0.0, lengths: [])

            var lastValue: Float = 0.0
            var lastY: Float = 0.0
            var first = true

            for pointValue in data ?? [] {
                guard let pointValue = pointValue as? NSValue else {
                    continue
                }
                let point = self.point(forValue: pointValue.cgPointValue)
                let x = Float(point.x)
                let y = Float(point.y)

                lastValue = Float(pointValue.cgPointValue.y)
                lastY = y

                if first {
                    context?.move(to: CGPoint(x: CGFloat(x), y: CGFloat(y)))
                    first = false

                    // Draw at least a dot if there's just one point
                    if (data?.count ?? 0) == 1 {
                        context?.addLine(to: CGPoint(x: CGFloat(Double(x) + VALUE_LINE_WIDTH), y: CGFloat(y)))
                    }
                } else {
                    context?.addLine(to: CGPoint(x: CGFloat(x), y: CGFloat(y)))
                }
            }

            if !first {

                // Draw line and label if there's at least one point
                context?.strokePath()

                // Draw last point label
                if let sysBoldFont = sysBoldFont {
                    attributesDict = [
                        kCTFontAttributeName : sysBoldFont,
                        kCTForegroundColorAttributeName : VALUE_LABEL_COLOR.cgColor
                    ]
                }

                let label = String(format: "%7.2f", lastValue)
                let labelAttr = NSAttributedString(string: label, attributes: attributesDict as? [NSAttributedString.Key : Any])
                context?.textPosition = CGPoint(x: CGFloat(Double(size.width) - RIGHT_MARGIN), y: CGFloat(lastY))
                line = CTLineCreateWithAttributedString(labelAttr as CFAttributedString)
                if let context = context {
                    CTLineDraw(line!, context)
                }
            }
        }

        // Cleanup


    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
}

let LEFT_AXIS_MARGIN = 40.0
let BOTTOM_AXIS_MARGIN = 20.0
let RIGHT_MARGIN = 40.0

let CHART_BACKGROUND_COLOR = UIColor(red: 0.7176, green: 0.8431, blue: 0.9647, alpha: 1.0000)
let AXIS_LINE_COLOR = UIColor.darkGray
let AXIS_LABEL_COLOR = UIColor.darkGray
let INTEGER_LINE_COLOR = UIColor.darkGray
let INTEGER_LABEL_COLOR = UIColor.darkGray
let VALUE_LINE_COLOR = UIColor.blue
let VALUE_LABEL_COLOR = UIColor.blue
let THRESHOLD_LINE_COLOR = UIColor.red
let THRESHOLD_LABEL_COLOR = UIColor.red

let AXIS_LINE_WIDTH = 2.0
let INTEGER_LINE_WIDTH = 1.0
let VALUE_LINE_WIDTH = 2.0
let THRESHOLD_LINE_WIDTH = 1.0

let FONT_SIZE = 11.0
